### PR TITLE
[IPO] NFC: avoid recalculating FunctionId hashes during ProfiledCallGraph construction

### PR DESCRIPTION
**Pull request by itrofimow**
This patch aims to reduce amount of time spent in recalculating `FunctionId` hashes (heavy md5) during `ProfiledCallGraph` construction

### CODE REVIEW

**Comment by itrofimow** [comment]
At my workplace we operate over huge binaries, which produce large sampling profiles (hundreds of MB in text format is not uncommon).

Compiling a no-op code
```
itrofimow@itrofimow-nix:~/work/profile_sample_use/clang_test$ cat main.cpp 
#include <iostream>

int main() {
	std::cout << "Hello!\n" << std::endl;

	return 0;
}
```
with such profiles takes
```
itrofimow@itrofimow-nix:~/work/profile_sample_use/clang_test$ time taskset -c 2 /home/itrofimow/work/llvm-project/clang_build/bin/clang++ -o main -O3 -fprofile-sample-use=/home/itrofimow/work/perfs/autofdo/15_09_2024/123MB.pgo.txt main.cpp

real	0m2.216s
user	0m2.084s
sys	0m0.132s
itrofimow@itrofimow-nix:~/work/profile_sample_use/clang_test$ time taskset -c 2 /home/itrofimow/work/llvm-project/clang_build/bin/clang++ -o main -O3 -fprofile-sample-use=/home/itrofimow/work/perfs/autofdo/15_09_2024/189MB.pgo.txt main.cpp

real	0m3.289s
user	0m3.068s
sys	0m0.214s
```
on current llvm main, and most of time time is spent in `ProfiledCallGraph` construction.

With this patch applied I'm seeing following numbers
```
itrofimow@itrofimow-nix:~/work/profile_sample_use/clang_test$ time taskset -c 2 /home/itrofimow/work/llvm-project/clang_build/bin/clang++ -o main -O3 -fprofile-sample-use=/home/itrofimow/work/perfs/autofdo/15_09_2024/123MB.pgo.txt main.cpp

real	0m1.753s
user	0m1.599s
sys	0m0.154s
itrofimow@itrofimow-nix:~/work/profile_sample_use/clang_test$ time taskset -c 2 /home/itrofimow/work/llvm-project/clang_build/bin/clang++ -o main -O3 -fprofile-sample-use=/home/itrofimow/work/perfs/autofdo/15_09_2024/189MB.pgo.txt main.cpp

real	0m2.620s
user	0m2.395s
sys	0m0.222s
```
which is ~20% speed up.

Please see the flamegraphs (warning: huge (15Mb) .svg files): [trunk](https://github.com/itrofimow/random_assets/blob/main/clang_main_123M.svg), [patch](https://github.com/itrofimow/random_assets/blob/main/clang_patch_123M.svg)

**Comment by itrofimow** [comment]
Hi @nikic! I see you are the one who recently reviewed a PR touching `ProfiledCallGraph.h`with the similar speedup idea, would you be able to give this PR a look?

**Review by nikic — COMMENTED** [review]
The idea here looks reasonable to me, but I'm not familiar with this code. (I see that FunctionId itself can also store a hash, so I'm a bit unsure if the solution here should be something else.)

File: llvm/include/llvm/Transforms/IPO/ProfiledCallGraph.h [hunk 1]
```diff
@@ -58,6 +58,19 @@ struct ProfiledCallGraphNode {
   edges Edges;
 };
 
+struct PrehashedFunctionId {
+
```
[the remark this answers is not in the document]
Comment (itrofimow) [reply llvm/include/llvm/Transforms/IPO/ProfiledCallGraph.h]: Done

File: llvm/include/llvm/Transforms/IPO/ProfiledCallGraph.h [hunk 2]
```diff
@@ -58,6 +58,19 @@ struct ProfiledCallGraphNode {
...
+  PrehashedFunctionId() = default;
+
+  /* implicit */ PrehashedFunctionId(FunctionId FId)
```
[the remark this answers is not in the document]
Comment (itrofimow) [reply llvm/include/llvm/Transforms/IPO/ProfiledCallGraph.h]: Done

**Comment by huangjd** [comment]
FunctionID is already capable of representing a hash value, so it is redundant to add another hash field. I think a more optimal way to do it is to do lazy hashing, whenever we put a function ID into the main sample profile map (or other maps), we convert it to hash representation. Although there's one complication here, that we will frequently need to compare strings coming from Function's name to FunctionID, so the Function object itself should also cache its name's  hash value.

**Comment by itrofimow** [comment]
Please note that I'm not adding another hash field, nor is it stored anywhere: for all intensive purposes `PrehashedFunctionId` is private to `ProfiledCallGraph` and doesn't leave its private scope.

W.r.t. to hashing `FunctionId`, `addProfiledCalls` currently does this (in pseudocode):
```
void addProfiledCalls(const FunctionSamples &Samples) {
    HashMap.lookup(Samples.GetFunction()); // from addProfiledFunction
    ....
    HashMap.lookup(Samples.GetFunction()); // from getBodySamples loop
    ....
    HashMap.lookup(Samples.GetFunction()); // from getCallsiteSamples loop
}
```
which calculates the hash in question 3 times. All this PR is doing is instead to calculate the hash once.

Although the `FunctionId` is capable to represent its hash value, it represents _either_ a string_view name or hash, and the name is used by callers in ways that are out of scope of my understanding of the code. I think the refactoring suggested could be implemented, but my knowledge wouldn't be sufficient to implement functional changes here, when this PR alone already brings very noticeable performance improvements.

I'll see if I can make the `PrehashesFunctionId` wrapper actually private to `ProfiledCallGraph` to avoid confusion.

**Pull request closed without merging**